robyn/openapi.py

```diff
@@ -0,0 +1,264 @@
+        """
+
+        query_params = None
+
+        if signature and "query_params" in signature.parameters:
+            query_params = signature.parameters["query_params"].default
+
+        modified_endpoint, path_obj = self.get_path_obj(endpoint, openapi_summary, openapi_tags, query_params)
+
+        if modified_endpoint not in self.openapi_spec["paths"]:
+            self.openapi_spec["paths"][modified_endpoint] = {}
+        self.openapi_spec["paths"][modified_endpoint][route_type] = path_obj
+
+    def add_subrouter_paths(self, subrouter_openapi):
+        """
+        Adds the subrouter paths to main router's openapi specs
+
+        :param subrouter_openapi: the OpenAPI object of the current subrouter
+        """
+        paths = subrouter_openapi.openapi_spec["paths"]
+
+        for path in paths:
+            self.openapi_spec["paths"][path] = paths[path]
+
+    def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None) -> (str, dict):
+        """
+        Get the "path" openapi object according to spec
+
+        :param endpoint: the endpoint to be added
+        :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
+        :param tags: tags -- for grouping of endpoints
+        :param query_params: query params for the function
+
+        :return: a tuple containing the endpoint with path params wrapped in braces and the "path" openapi object
+        according to spec
+        """
+        # robyn has paths like /:url/:etc whereas openapi requires path like /{url}/{path}
+        # this function is used for converting path params to the required form
+        # initialized with endpoint for handling endpoints without path params
+        endpoint_with_path_params_wrapped_in_braces = endpoint
+
+        endpoint_path_params_split = endpoint.split(":")
+        openapi_parameter_object = []
+
+        if len(endpoint_path_params_split) > 1:
+            endpoint_without_path_params = endpoint_path_params_split[0]
+
+            endpoint_with_path_params_wrapped_in_braces = (
+                endpoint_without_path_params[:-1] if endpoint_without_path_params.endswith("/") else endpoint_without_path_params
+            )
+
+            for path_param in endpoint_path_params_split[1:]:
+                path_param_name = path_param[:-1] if path_param.endswith("/") else path_param
+
+                openapi_parameter_object.append(
+                    {
+                        "name": path_param_name,
+                        "in": "path",
+                        "required": True,
+                        "schema": {"type": "string"},
+                    }
+                )
+                endpoint_with_path_params_wrapped_in_braces += "/{" + path_param_name + "}"
+
+        if query_params:
+            for query_param in query_params.__annotations__:
```

Comment:
You can use `get_type_hints` form `typing`.

I used AI for some suggestion as I not at home atm. But this sample can come in hand

```
from typing import get_type_hints, TypedDict, Union

def get_path_obj(self, endpoint: str, summary: str, tags: list, query_params: TypedDict = None) -> (str, dict):
    """
    Get the "path" openapi object according to spec

    :param endpoint: the endpoint to be added
    :param summary: short summary of the endpoint (to be fetched from the endpoint defenition by default)
    :param tags: tags -- for grouping of endpoints
    :param query_params: query params for the function

    :return: a tuple containing the endpoint with path params wrapped in braces and the "path" openapi object
    according to spec
    """
    # robyn has paths like /:url/:etc whereas openapi requires path like /{url}/{path}
    # this function is used for converting path params to the required form
    # initialized with endpoint for handling endpoints without path params
    endpoint_with_path_params_wrapped_in_braces = endpoint

    endpoint_path_params_split = endpoint.split(":")
    openapi_parameter_object = []

    if len(endpoint_path_params_split) > 1:
        endpoint_without_path_params = endpoint_path_params_split[0]

        endpoint_with_path_params_wrapped_in_braces = (
            endpoint_without_path_params[:-1] if endpoint_without_path_params.endswith("/") else endpoint_without_path_params
        )

        for path_param in endpoint_path_params_split[1:]:
            path_param_name = path_param[:-1] if path_param.endswith("/") else path_param

            openapi_parameter_object.append(
                {
                    "name": path_param_name,
                    "in": "path",
                    "required": True,
                    "schema": {"type": "string"},
                }
            )
            endpoint_with_path_params_wrapped_in_braces += "/{" + path_param_name + "}"

    if query_params:
        # Use get_type_hints to retrieve the actual types from the annotations
        query_param_types = get_type_hints(query_params)
        
        for query_param, query_param_type in query_param_types.items():
            openapi_type, custom_type = self.get_openapi_type(query_param_type)

            param_schema = {"type": openapi_type}
            if custom_type:
                param_schema["$ref"] = f"#/components/schemas/{custom_type}"

            openapi_parameter_object.append(
                {
                    "name": query_param,
                    "in": "query",
                    "required": False,
                    "schema": param_schema,
                }
            )

    return endpoint_with_path_params_wrapped_in_braces, {
        "summary": summary,
        "tags": tags,
        "parameters": openapi_parameter_object,
        "responses": {
            "200": {
                "description": "Successful Response",
            }
        },
    }

def get_openapi_type(self, python_type: Union[type, str]) -> (str, Optional[str]):
    """
    Converts Python types to OpenAPI types, handling custom types.

    :param python_type: The Python type to be converted
    :return: A tuple with OpenAPI type as a string and custom type name if applicable
    """
    type_mapping = {
        int: "integer",
        str: "string",
        bool: "boolean",
        float: "number",
        dict: "object",
        list: "array",
        # Add more mappings as necessary
    }

    # If it's a standard type, return the OpenAPI type directly
    if python_type in type_mapping:
        return type_mapping[python_type], None

    # If it's a custom type (likely a class), return its name for the $ref
    if hasattr(python_type, '__name__'):
        return "object", python_type.__name__

    # Handle cases where python_type is provided as a string
    if isinstance(python_type, str):
        python_type = eval(python_type)
        if python_type in type_mapping:
            return type_mapping[python_type], None
        elif hasattr(python_type, '__name__'):
            return "object", python_type.__name__

    # Default to "string" if type is not found
    return "string", None

```